File: maps/src/main/java/it/localhostsoftware/maps/huawei/model/HuaweiMarker.java
package it.localhostsoftware.maps.huawei.model;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import it.localhostsoftware.maps.model.BitmapDescriptor;
import it.localhostsoftware.maps.model.LatLng;
import it.localhostsoftware.maps.model.Marker;

public class HuaweiMarker extends Marker<com.huawei.hms.maps.model.Marker> {
    public HuaweiMarker(com.huawei.hms.maps.model.Marker marker) {
        super(marker);
    }

    @Override
    public void remove() {
        getMarker().remove();
    }

    @Override
    public String getId() {
        return getMarker().getId();
    }

    @Override
    public void setPosition(@NonNull LatLng<?> latLng) {
        getMarker().setPosition((com.huawei.hms.maps.model.LatLng) latLng.getLatLng());
    }

    @Override
    public LatLng<?> getPosition() {
        return new HuaweiLatLng(getMarker().getPosition());
    }

    @Override
    public void setZIndex(float v) {
        getMarker().setZIndex(v);
    }

    @Override
    public float getZIndex() {
        return getMarker().getZIndex();
    }

    @Override
    public void setIcon(@Nullable BitmapDescriptor<?> bitmapDescriptor) {
        getMarker().setIcon((com.huawei.hms.maps.model.BitmapDescriptor) bitmapDescriptor.getBitmapDescriptor());
    }

    @Override
    public void setAnchor(float v, float v1) {
        getMarker().setMarkerAnchor(v, v1);
    }

    @Override
    public void setInfoWindowAnchor(float v, float v1) {
        getMarker().setInfoWindowAnchor(v, v1);
    }

    @Override
    public void setTitle(@Nullable String s) {
        getMarker().setTitle(s);
    }

    @Override
    public String getTitle() {
        return getMarker().getTitle();
    }

    @Override
    public void setSnippet(@Nullable String s) {
        getMarker().setSnippet(s);
    }

    @Override
    public String getSnippet() {
        return getMarker().getSnippet();
    }

    @Override
    public void setDraggable(boolean b) {
        getMarker().setDraggable(b);
    }

    @Override
    public boolean isDraggable() {
        return getMarker().isDraggable();
    }

    @Override
    public void showInfoWindow() {
        getMarker().showInfoWindow();
    }

    @Override
    public void hideInfoWindow() {
        getMarker().hideInfoWindow();
    }

    @Override
    public boolean isInfoWindowShown() {
        return getMarker().isInfoWindowShown();
    }

    @Override
    public void setVisible(boolean b) {
        getMarker().setVisible(b);
    }

    @Override
    public boolean isVisible() {
        return getMarker().isVisible();
    }

    @Override
    public void setFlat(boolean b) {
        getMarker().setFlat(b);
    }

    @Override
    public boolean isFlat() {
        return getMarker().isFlat();
    }

    @Override
    public void setRotation(float v) {
        getMarker().setRotation(v);
    }

    @Override
    public float getRotation() {
        return getMarker().getRotation();
    }

    @Override
    public void setAlpha(float v) {
        getMarker().setAlpha(v);
    }

    @Override
    public float getAlpha() {
        return getMarker().getAlpha();
    }

    @Override
    public void setTag(@Nullable Object o) {
        getMarker().setTag(o);
    }

    @Nullable
    @Override
    public Object getTag() {
        return getMarker().getTag();
    }
}

File: maps/src/main/kotlin/it/localhostsoftware/maps/huawei/model/HuaweiMarker.kt
package it.localhostsoftware.maps.huawei.model

import com.huawei.hms.maps.model.Marker
import com.huawei.hms.maps.model.MarkerOptions

class HuaweiMarker(marker: Marker) : it.localhostsoftware.maps.model.Marker<Marker, HuaweiLatLng, HuaweiBitmapDescriptor>(marker) {
    override fun remove() {
        m.remove()
    }

    override val id: String
        get() = m.id
    override var position: HuaweiLatLng
        get() = HuaweiLatLng(m.position)
        set(value) {
            m.position = value.ll
        }
    override var zIndex: Float
        get() = m.zIndex
        set(v) {
            m.zIndex = v
        }

    override fun setIcon(var1: HuaweiBitmapDescriptor?) {
        m.setIcon(var1?.bitmapDescriptor)
    }

    override fun setAnchor(var1: Float, var2: Float) {
        m.setMarkerAnchor(var1, var2)
    }

    override fun setInfoWindowAnchor(var1: Float, var2: Float) {
        m.setInfoWindowAnchor(var1, var2)
    }

    override var title: String?
        get() = m.title
        set(s) {
            m.title = s
        }
    override var snippet: String?
        get() = m.snippet
        set(s) {
            m.snippet = s
        }
    override var isDraggable: Boolean
        get() = m.isDraggable
        set(b) {
            m.isDraggable = b
        }

    override fun showInfoWindow() {
        m.showInfoWindow()
    }

    override fun hideInfoWindow() {
        m.hideInfoWindow()
    }

    override val isInfoWindowShown: Boolean
        get() = m.isInfoWindowShown
    override var isVisible: Boolean
        get() = m.isVisible
        set(b) {
            m.isVisible = b
        }
    override var isFlat: Boolean
        get() = m.isFlat
        set(b) {
            m.isFlat = b
        }
    override var rotation: Float
        get() = m.rotation
        set(v) {
            m.rotation = v
        }
    override var alpha: Float
        get() = m.alpha
        set(v) {
            m.alpha = v
        }
    override var tag: Any?
        get() = m.tag
        set(o) {
            m.tag = o
        }
}

class HuaweiMarkerOptions(markerOptions: MarkerOptions) : it.localhostsoftware.maps.model.MarkerOptions<MarkerOptions, HuaweiBitmapDescriptor, HuaweiLatLng>(markerOptions) {
    override var position: HuaweiLatLng
        get() = HuaweiLatLng(mo.position)
        set(value) {
            mo.position(value.ll)
        }
    override var title: String?
        get() = mo.title
        set(value) {
            mo.title(value)
        }
    override var snippet: String?
        get() = mo.snippet
        set(value) {
            mo.snippet(value)
        }
    override var icon: HuaweiBitmapDescriptor?
        get() = mo.icon?.let { HuaweiBitmapDescriptor(it) }
        set(value) {
            mo.icon(value?.bitmapDescriptor)
        }
    override var isDraggable: Boolean
        get() = mo.isDraggable
        set(value) {
            mo.draggable(value)
        }
    override var isVisible: Boolean
        get() = mo.isVisible
        set(value) {
            mo.visible(value)
        }
    override var isFlat: Boolean
        get() = mo.isFlat
        set(value) {
            mo.flat(value)
        }
    override var rotation: Float
        get() = mo.rotation
        set(value) {
            mo.rotation(value)
        }
    override var alpha: Float
        get() = mo.alpha
        set(value) {
            mo.alpha(value)
        }
    override var zIndex: Float
        get() = mo.zIndex
        set(value) {
            mo.zIndex(value)
        }
    override var anchor: Pair<Float, Float>
        get() = Pair(mo.markerAnchorU, mo.markerAnchorV)
        set(value) {
            mo.anchorMarker(value.first, value.second)
        }
    override var infoWindowAnchor: Pair<Float, Float>
        get() = Pair(mo.infoWindowAnchorU, mo.infoWindowAnchorV)
        set(value) {
            mo.infoWindowAnchor(value.first, value.second)
        }
}